test/e2e/test_cluster_configs/byok.json

```diff
@@ -0,0 +1,42 @@
+{
+	"env": {
```

Comment:
@devigned is this config all for such an e2e test? do I need to configure sth else?

Reply:
Yes, this is all that’s needed for an Jenkins e2e test. Jenkins won’t run this until a job is triggered or created to run this branch / fork. Once in master, it’ll run periodically. Perhaps work with @jackfrancis tomorrow to create that job to verify this config.

Reply:
FYI, I've added a `INFRA_RESOURCE_GROUP` env var to the E2E template generation processor, and we're injecting that and the `SUBSCRIPTION_ID` env var into the `diskEncryptionSetID` property via string replacement, so we can formulate the URI per-test via env vars.

Reply:
@jackfrancis thanks, I have configured a `diskEncryptionSet` and added into e2e test config:
```
"/subscriptions/SUB_ID/resourceGroups/RESOURCE_GROUP/providers/Microsoft.Compute/diskEncryptionSets/des"
```

Reply:
I've put together a one-off that successfully deploys a cluster using a VMSS pool w/ `diskEncryptionSetID`. How do we actually test this functionality from the Kubernetes layer? How can I prove that the OS disk on these VMSS instances have the desired configuration?

Reply:
@jackfrancis
good question, here is the full E2e test, already verified on an aks-engine cluster:
```
### E2E test
1. Create a cluster with BYOK feature enabled
2. Make sure all agent nodes are up
3. Revoke DiskEncryptionSet access to azure key vault
4. After a few minutes, make sure all agent nodes are down
5. Grant DiskEncryptionSet access to azure key vault
6. After a few minutes, make sure all agent nodes are up
```

Not sure whether aks-engine test CI supports such testing scenario, and also need to "translate" below DiskEncryptionSet related script to golang code if we want to write a fully automatic e2e test:
https://github.com/andyzhangx/demo/blob/master/aks/byok/create-diskencryptionset.sh

Shall we go with this PR first? Thanks.

Reply:
> 3. Revoke DiskEncryptionSet access to azure key vault

How do you do this?

> 5. Grant DiskEncryptionSet access to azure key vault

How do you do this?

Reply:
@jackfrancis by be

> > 1. Revoke DiskEncryptionSet access to azure key vault
> 
> How do you do this?
> 
> > 1. Grant DiskEncryptionSet access to azure key vault
> 
> How do you do this?

@jackfrancis by below command in https://github.com/andyzhangx/demo/blob/master/aks/byok/create-diskencryptionset.sh
```
az role assignment create --assignee $desIdentity --role Reader --scope $keyVaultId
az role assignment delete --assignee $desIdentity --role Reader
```
How can we support such a test case? Is there already an example?

Reply:
I think what @jackfrancis was really trying to get at is whether or not DES will be observable from Kubernetes. Would some process running on the node be able to tell that the underlying infrastructure has DES enabled without asking Azure Resource Manager? Is this feature completely transparent to the node? Could someone check `blkid` or something to see type of block device type, or is it at rest encryption and only known by Azure?

Reply:
> I think what @jackfrancis was really trying to get at is whether or not DES will be observable from Kubernetes. Would some process running on the node be able to tell that the underlying infrastructure has DES enabled without asking Azure Resource Manager? Is this feature completely transparent to the node? Could someone check `blkid` or something to see type of block device type, or is it at rest encryption and only known by Azure?

@devigned it's a [server side encryption](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/disk-encryption), not Azure Disk Encryption, `Azure Disk Encryption` is another feature:
"
Azure Disk Encryption leverages the BitLocker feature of Windows and the DM-Crypt feature of Linux to encrypt managed disks with customer-managed keys within the guest VM. Server-side encryption with customer-managed keys improves on ADE by enabling you to use any OS types and images for your VMs by encrypting data in the Storage service.
"
details: https://docs.microsoft.com/en-us/azure/virtual-machines/windows/disk-encryption#server-side-encryption-versus-azure-disk-encryption

Reply:
Yep. That is what I had figured. The VM should have no idea that this feature is enabled. 

The extent that we can test this seems to be Azure accepts the deployment template, provisions the infra with configured data encryption, and everything runs as expected.

Reply:
> Yep. That is what I had figured. The VM should have no idea that this feature is enabled.
> 
> The extent that we can test this seems to be Azure accepts the deployment template, provisions the infra with configured data encryption, and everything runs as expected.

so shall we merge this PR now? Thanks.

Reply:
O.K., we have basic, working E2E coverage of the ARM deployment + BYOK scenario. That will do for now.